Platform/AMD/AmdPlatformPkg/DynamicTables/Library/Acpi/AcpiSsdtPciLib/AcpiSsdtPciLib.c

```diff
@@ -0,0 +1,1540 @@
+               TRUE,
+               CrsNode,
+               NULL
+               );
+
+    Status = AmlCodeGenRdQWordMemory (
+               FALSE,
+               TRUE,
+               TRUE,
+               TRUE,
+               FALSE, // non cacheable
+               TRUE,
+               0,
+               0xFEDCC000,
+               0xFEDFFFFF,
+               0x0,
+               0x34000,
+               0,
+               NULL,
+               0,
+               TRUE,
+               CrsNode,
+               NULL
+               );
+
+    Status = AmlCodeGenRdQWordMemory (
+               FALSE,
+               TRUE,
+               TRUE,
+               TRUE,
+               FALSE, // non cacheable
+               TRUE,
+               0,
+               0xFEE01000,
+               0xFEFFFFFF,
+               0x0,
+               0x1FF000,
+               0,
+               NULL,
+               0,
+               TRUE,
+               CrsNode,
+               NULL
+               );
+  }
+
+  if (EFI_ERROR (Status)) {
+    Status = EFI_DEVICE_ERROR;
+    return Status;
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Insert Root Port into the AML table
+
+  @param[in]      RootBridge  - Single Root Bridge instance
+  @param[in]      GlobalInterruptBase - Base to add to IOAPIC interrupt offset
+  @param[in,out]  PciNode    - AmlLib table node
+
+  @retval         EFI_SUCCESS, various EFI FAILUREs.
+**/
+EFI_STATUS
+EFIAPI
+InternalInsertRootPorts (
+  IN      AMD_PCI_ROOT_BRIDGE_OBJECT_INSTANCE  *RootBridge,
+  IN      UINTN                                GlobalInterruptBase,
+  IN OUT  AML_OBJECT_NODE_HANDLE               PciNode
+  )
+{
+  AML_METHOD_PARAM        MethodParam[7];
+  AML_OBJECT_NODE_HANDLE  DeviceNode;
+  AML_OBJECT_NODE_HANDLE  DsmMethod;
+  AML_OBJECT_NODE_HANDLE  OstMethod;
+  AML_OBJECT_NODE_HANDLE  PrtNode;
+  CHAR8                   NameSeg[5];
+  CHAR8                   RpName[15];
+  EFI_STATUS              Status;
+  UINTN                   Index;
+  UINTN                   RPIndex;
+
+  for (RPIndex = 1; RPIndex <= RootBridge->RootPortCount; RPIndex++) {
+    if ((RootBridge->RootPort[RPIndex]->PortPresent == 0) &&
+        (RootBridge->RootPort[RPIndex]->Enabled == 0))
+    {
+      continue;
+    }
+
+    CopyMem (NameSeg, "RPxx", AML_NAME_SEG_SIZE + 1);
+    NameSeg[AML_NAME_SEG_SIZE - 2] = AsciiFromHex (RootBridge->RootPort[RPIndex]->Device & 0xF);
+    NameSeg[AML_NAME_SEG_SIZE - 1] = AsciiFromHex (RootBridge->RootPort[RPIndex]->Function & 0xF);
+
+    Status = AmlCodeGenDevice (NameSeg, PciNode, &DeviceNode);
+    if (EFI_ERROR (Status)) {
+      ASSERT_EFI_ERROR (Status);
+      return Status;
+    }
+
+    Status = AmlCodeGenNameInteger (
+               "_ADR",
+               (RootBridge->RootPort[RPIndex]->Device << 16) + RootBridge->RootPort[RPIndex]->Function,
+               DeviceNode,
+               NULL
+               );
+    if (EFI_ERROR (Status)) {
+      ASSERT_EFI_ERROR (Status);

```

Comment:
Do we have to delete the newly created node (DeviceNode) in the fail conditions in this file?

Reply:
done, added the exit_handler